### DISCUSSION TITLE
Misbeavings with target x86_64-unknown-linux-musl

### DISCUSSION THREAD

**Discussion by lanceschi** [helix-editor/helix#10356] [Troubleshooting]
### Reproduction steps
I managed to build [helix editor][helix-editor-url] v0.6.0 with target set to `x86_64-unknown-linux-musl`.

Here's the [Dockerfile][dockerfile-url]. If I extract the binary from the built image and I run it on my Ubuntu 20.04.3 LTS desktop for instance, it works as expected. It did not under the Alpine Linux v3.15.0 container. The theme doesn't seem to be applied as well as the syntax highlighting.

Build the above [Dockerfile][dockerfile-url]:
```shell
docker build \
  --file ./Dockerfile \
  --pull -t "<DOCKER-IMAGE-NAME>" .
```

... and run the container with the following:
```shell
docker run --rm -it <DOCKER-IMAGE-NAME> /bin/ash
```
From within the running container download an example file:
```shell
wget https://raw.githubusercontent.com/lanceschi/threejs-ply-loader/develop/tests/index.spec.js
```
Open it verbosely:
```shell
./hx -vvv index.spec.js
```
Here's a screenshot:
![helix-editor-alpine](https://user-images.githubusercontent.com/3028034/151674880-581448a0-b9e7-4b26-8ad8-88eb67fa976a.jpg)

Here the content of Alpine Linux file `~/.cache/helix/helix.log`:
```shell
2022-01-29T19:28:27.724 helix_core::syntax [INFO] Error opening dynamic library "/helix-v0.6.0/runtime/grammars/javascript.so"
2022-01-29T19:28:27.725 helix_view::editor [ERROR] Failed to initialize the LSP for `source.js` { LSP not defined }
2022-01-29T19:28:27.725 mio::poll [TRACE] registering event source with poller: token=Token(1), interests=READABLE | WRITABLE
2022-01-29T19:28:27.725 mio::poll [TRACE] registering event source with poller: token=Token(2), interests=READABLE | WRITABLE
2022-01-29T19:28:27.725 mio::poll [TRACE] registering event source with poller: token=Token(0), interests=READABLE
2022-01-29T19:28:27.725 mio::poll [TRACE] registering event source with poller: token=Token(1), interests=READABLE
2022-01-29T19:30:05.388 mio::poll [TRACE] deregistering event source from poller
2022-01-29T19:30:05.388 mio::poll [TRACE] deregistering event source from poller
```
Here's the screenshot of the same file opened with the same `musl` binary on the host Ubuntu 20.04.3 LTS:

![helix-ubuntu](https://user-images.githubusercontent.com/3028034/151675193-a996fe41-4f5a-4528-984a-c997b10dedce.jpg)

Here's the Ubuntu log `~/.cache/helix/helix.log`:
```shell
2022-01-29T20:41:31.821 helix_core::syntax [INFO] Error opening dynamic library "/home/luca/dev/tools-in-rust/helix-editor-musl2/dist/helix-v0.6.0/runtime/grammars/javascript.so"
2022-01-29T20:41:31.822 helix_view::editor [ERROR] Failed to initialize the LSP for `source.js` { LSP not defined }
2022-01-29T20:41:31.822 mio::poll [TRACE] registering event source with poller: token=Token(1), interests=READABLE | WRITABLE
2022-01-29T20:41:31.822 mio::poll [TRACE] registering event source with poller: token=Token(2), interests=READABLE | WRITABLE
2022-01-29T20:41:31.822 mio::poll [TRACE] registering event source with poller: token=Token(0), interests=READABLE
2022-01-29T20:41:31.822 mio::poll [TRACE] registering event source with poller: token=Token(1), interests=READABLE
2022-01-29T20:41:33.929 mio::poll [TRACE] deregistering event source from poller
2022-01-29T20:41:33.929 mio::poll [TRACE] deregistering event source from poller

```

### Environment

- Platform: Alpine Linux v3.15.0
- Terminal emulator: [Terminator][terminator-url] (and [WezTerm][wezterm-url])
- Helix version: helix v0.6.0-dirty

Thanks :100: for the awesome app!!!

[helix-editor-url]: https://github.com/helix-editor/helix/tree/v0.6.0
[dockerfile-url]: https://github.com/lanceschi/helix-editor-musl/blob/feature/scaffolding/Dockerfile
[terminator-url]: https://terminator-gtk3.readthedocs.io/en/latest/
[wezterm-url]: https://github.com/wez/wezterm

**Comment by dead10ck** [comment]
You have to tell helix where to find the runtime directory, so you'll need something like

```dockerfile
ENV HELIX_RUNTIME="/helix-${VERSION}/runtime"
```

**Comment by archseer** [comment]
Ah cool, you got it to build on musl? Linking back to an older thread where there was some issues #1028

**Comment by lanceschi** [comment]
> You have to tell helix where to find the runtime directory, so you'll need something like
> 
> ```dockerfile
> ENV HELIX_RUNTIME="/helix-${VERSION}/runtime"
> ```
Hello @dead10ck, and thanks for your advice. I made some minor edits to the [Dockerfile][dockerfile-url] and I added the following line:
```dockerfile
ENV PATH=/helix-${VERSION}:"$PATH"
```
... so that the helix editor can be invoked simply issuing `hx` without the `./` folder prefix. No noticeable difference though.
I tried your suggestion within the live running container issuing the following:
```shell
export HELIX_RUNTIME="/helix-${VERSION}/runtime"
```
Again no noticeable difference:
![helix-alpine](https://user-images.githubusercontent.com/3028034/151704106-9e1400d0-d478-44fd-90ec-713806742651.jpg)

[dockerfile-url]: https://github.com/lanceschi/helix-editor-musl/blob/feature/scaffolding/Dockerfile

**Comment by dead10ck** [comment]
Ah I see, maybe it's related to #1028 then.

**Comment by lanceschi** [comment]
> Ah cool, you got it to build on musl? Linking back to an older thread where there was some issues #1028

Hello @archseer,

yes, I setup a [Dockerfile][dockerfile-url] in a [WIP](https://en.wikipedia.org/wiki/Work_in_process "Work In Progress") github [repo][repo-url].
You can build the docker image yourself or pull a prebuilt image:

```shell
docker pull lanceschi/helix-editor-musl:latest
```
...  and have a look:
```shell
docker run --rm -it lanceschi/helix-editor-musl:latest /bin/ash
```
You can also have a look at the [logs][log-url] of the Github Dockerfile build action.

[dockerfile-url]: https://github.com/lanceschi/helix-editor-musl/blob/feature/scaffolding/Dockerfile
[repo-url]: https://github.com/lanceschi/helix-editor-musl/tree/feature/scaffolding
[log-url]: https://github.com/lanceschi/helix-editor-musl/runs/4997356987?check_suite_focus=true

**Comment by jirutka** [comment]
I’ve packaged Helix in Alpine Linux (package [helix](https://pkgs.alpinelinux.org/packages?name=helix&branch=edge)), so you don’t have to build it yourself anymore. It’s currently in the testing repo, I’m gonna move it to the community repo soon.

**Comment by lanceschi** [comment]
Hello @jirutka,

thanks a lot! I had a try issuing the following from a `alpine:3.15.0` base image:
```shell
echo "@testing http://dl-cdn.alpinelinux.org/alpine/edge/testing" >> /etc/apk/repositories
apk add helix@testing
```
The post message install note reports:
```shell
fetch http://dl-cdn.alpinelinux.org/alpine/edge/testing/x86_64/APKINDEX.tar.gz
(1/1) Installing helix@testing (0.6.0-r0)
Executing helix-0.6.0-r0.post-install
*
* Tree-sitter grammars for syntax highlighting are packaged separately.
* Install either selected tree-sitter-<lang> packages for the languages you
* use or a "virtual package" tree-sitter-grammars to install all of them.
*
```
How am I supposed to install those?
Maybe that's why I can't get syntax highlighting?

Another thing I noticed checking the helix version number:
```shell
$ hx -V
helix v3.15.0-5432-gd33618c49f
```
Isn't it supposed to be `helix v0.6.0`?

**Comment by jirutka** [comment]
> How am I supposed to install those?

I don’t understand what’s unclear on this. They are packaged separately, i.e. there are packages [tree-sitter-rust](https://pkgs.alpinelinux.org/packages?name=tree-sitter-rust), [tree-sitter-ruby](https://pkgs.alpinelinux.org/packages?name=tree-sitter-ruby) etc. which you can install to get syntax highlighting for the given languages. Or you can install `tree-sitter-grammars` which will install all tree-sitter grammar packages available in the Alpine repository. Packages are installed using `apk add`, as you already did to install helix itself.

> $ hx -V
> helix v3.15.0-5432-gd33618c49f

Thanks, I’ll fix this.
The problem here is that helix foolishly reads its own version during the build from a git repository where the sources are unpacked. This version (v3.15.0-5432-gd33618c49f) actually comes from our aports repository, where build recipes are stored and packages are built from.

**Comment by archseer** [comment]
@jirutka Thank you for packaging helix! 🎉 

> The problem here is that helix foolishly reads its own version during the build from a git repository where the sources are unpacked.

Yeah this was my concern when the change was introduced in https://github.com/helix-editor/helix/pull/957. I'll revert that and only pull the commit sha from the repository. That won't fully fix it but at least the version will be correct.

EDIT: https://github.com/helix-editor/helix/commit/d3221b03a2b3ee5acc68066f34c4e56e2d05ce94

**Comment by lanceschi** [comment]
Thanks :100: for the clarification @jirutka. I followed your advices:
```shell
echo "@edge https://dl-cdn.alpinelinux.org/alpine/edge/community/" >> /etc/apk/repositories
apk add tree-sitter-javascript@edge
```
Now I gained syntax highlighting for `JavaScript` files, still the helix default purple theme doesn't seem to be applied:
![alpine-helix](https://user-images.githubusercontent.com/3028034/152193177-5042d2bf-91c6-4592-92b0-fbe1bf405a54.jpg)


Any clues?

**Comment by archseer** [comment]
What terminal are you using? Does it support true color?

**Comment by lanceschi** [comment]
@archseer I tried with both [Terminator][terminator-url] (and [WezTerm][wezterm-url])

I tried with both the more advanced [bash][bash-url] as well as [fish][fish-url] Unix shells instead of the default [ash][ash-url] that comes with Alpine Linux.

Please also consider that both terminals works good with my host computer.

[terminator-url]: https://terminator-gtk3.readthedocs.io/en/latest/
[wezterm-url]: https://github.com/wez/wezterm
[bash-url]: https://en.wikipedia.org/wiki/Bash_(Unix_shell)
[fish-url]: https://github.com/fish-shell/fish-shell
[ash-url]: https://en.wikipedia.org/wiki/Almquist_shell

**Comment by jirutka** [comment]
> Thank you for packaging helix! tada

You’re welcome!

> Yeah this was my concern when the change was introduced in #957. I'll revert that and only pull the commit sha from the repository. That won't fully fix it but at least the version will be correct.
>
> EDIT: d3221b0

You can test if there’s a `.git` directory in the root of the project; if it’s not, skip pulling any version information from git.

**Comment by jirutka** [comment]
I can confirm @archseer’s findings, I also see black background. I didn’t know it was supposed to be purple by default, so I didn’t initially consider it a bug. However, I dunno what can be the cause.

**Comment by jirutka** [comment]
I just tried it on my Alpine desktop in [Kitty](https://sw.kovidgoyal.net/kitty/) terminal and I see a purple background.

Relevant env. variables:

```
COLORTERM=truecolor
TERM=xterm-kitty
```

**Comment by lanceschi** [comment]
`export COLORTERM=truecolor` did the trick for me. Thanks @jirutka

**Comment by archseer** [comment]
Weird, wezterm should be already setting the COLORTERM var.

**Comment by lanceschi** [comment]
> Weird, wezterm should be already setting the COLORTERM var.

No, actually it makes sense @archseer. I launch [WezTerm][wezterm-url] from the host computer where `COLORTERM` is correctly set to `truecolor`, whereas in the running Alpine container that specific ENV variable turns out to be unset.

I faced the same issue with a [CentOS v7.x][centos-url] powered machine as well.

[wezterm-url]: https://wezfurlong.org/wezterm/
[centos-url]: https://www.centos.org/

**Comment by lanceschi** [comment]
@jirutka is it possible to see the Dockerfile or CLI command sequence you're using to build helix?

**Comment by jirutka** [comment]
[testing/helix](https://gitlab.alpinelinux.org/alpine/aports/-/tree/4995bdfc4067f21b2ba8873bb5803bdc47bb38d9/testing/helix) in the Alpine aports repository. However, you will not find anything special here. Maybe the key difference is that you build it with the stock rustc instead of rustc from Alpine package.
